src/Terrain/Maps/MapReader.java

```diff
@@ -0,0 +1,171 @@
+      { SE, FC, FC, FC, GR, CT, SE, GR, CT, SE }, // 13
+      { SE, SE, SE, SE, SE, SE, SE, SE, SE, SE } };// 14
+  private static XYCoord[] co1Props = { new XYCoord(1, 5), new XYCoord(1, 6), new XYCoord(2, 6), new XYCoord(2, 7),
+      new XYCoord(3, 7) };
+  private static XYCoord[] co2Props = { new XYCoord(11, 1), new XYCoord(12, 2), new XYCoord(13, 1), new XYCoord(13, 2),
+      new XYCoord(13, 3) };
+  private static XYCoord[][] properties = { co1Props, co2Props };
+  private static MapInfo info = new MapInfo(mapName, terrainData, properties);
+
+  /**
+   * Tells the MapReader to read in the maps.
+   */
+  public static ArrayList<MapInfo> readMapData()
+  {
+    ArrayList<MapInfo> importMaps = new ArrayList<MapInfo>();
+
+    try
+    {
+      final File folder = new File("res/map");
+
+      for( final File fileEntry : folder.listFiles() )
+      {
+        // we aren't checking subdirectories, yet
+        if( !fileEntry.isDirectory() )
+        {
+          try
+          {
+            // System.out.println();
+            String mapName = fileEntry.getName();
+            if( !mapName.contains(".map") )
+              continue;
+            mapName = mapName.replaceAll("_", " ");
+            mapName = mapName.replaceAll(".map", "");
+            System.out.println("Parsing map: " + mapName);
+            Scanner scanner = new Scanner(fileEntry);
+            String line = scanner.nextLine();
+            ArrayList<ArrayList<Environment.Terrains>> terrainData = new ArrayList<ArrayList<Environment.Terrains>>();
+
+            ArrayList<ArrayList<XYCoord>> properties = new ArrayList<ArrayList<XYCoord>>(4);
+            properties.add(new ArrayList<XYCoord>());
+            properties.add(new ArrayList<XYCoord>());
+            properties.add(new ArrayList<XYCoord>());
+            properties.add(new ArrayList<XYCoord>());
+
+            for( int i = 0; i < line.length() / 4; i++ )
+            {
+              terrainData.add(new ArrayList<Environment.Terrains>());
+            }
+            for( int yCoord = 0; scanner.hasNextLine(); line = scanner.nextLine() )
+            {
+              for( int i = 0; i < line.length(); i += 4 )
+              {
+                int sideNumber = 99;
+                try
+                {
+                  sideNumber = Integer.parseInt(("" + line.charAt(i) + line.charAt(i + 1)).trim());
+                }
+                catch (NumberFormatException e)
+                {}
+                if( sideNumber < 4 ) // TODO: add more sides
+                {
+                  properties.get(sideNumber).add(new XYCoord(i / 4, yCoord));
+                }
+                String terrainCode = "" + line.charAt(i + 2) + line.charAt(i + 3);
+                terrainData.get(i / 4).add(stringToCode(terrainCode));
+              }
+              yCoord++;
+            }
+            ArrayList<Environment.Terrains[]> terrainArrayArray = new ArrayList<Environment.Terrains[]>();
```

Comment:
Is there a reason to create `terrainArrayArray` as an intermediate before calling `toArray` on it? Can we just go to a 2D array in one step instead of taking two (and copying everything twice)? Same with `propertyArrayArray`.

If we could find a way to avoid the need to do this conversion alltogether, that would be even better, but I'm not sure we can reasonably avoid that.

Reply:
I looked into it some while I was doing it, and didn't find anything useful.